File: src/edu/uf/interactable/Transferrin.java
package edu.uf.interactable;
import java.util.HashMap;
import java.util.Map;

import edu.uf.Diffusion.Diffuse;
import edu.uf.compartments.GridFactory;
import edu.uf.intracellularState.AspergillusMacrophage;
import edu.uf.intracellularState.FMacrophageBooleanNetwork;
import edu.uf.primitives.Interactions;
import edu.uf.utils.Constants;
import edu.uf.utils.Util;


public class Transferrin extends Molecule{
    

	public static final String NAME = "Transferrin";
	public static final int NUM_STATES = 3;
    private static final double THRESHOLD = Constants.K_M_TF_TAFC *Constants.VOXEL_VOL / 1.0e6;
    
    private static Transferrin molecule = null;

    private static final Map<String, Integer> INDEXES;
    
    static {
    	INDEXES = new HashMap<>();
    	INDEXES.put("Tf", 0);
    	INDEXES.put("TfFe", 1); 
    	INDEXES.put("TfFe2", 2);
    }
    
    private Transferrin(double[][][][] qttys, Diffuse diffuse) {
		super(qttys, diffuse);
	}
    
    public static Transferrin getMolecule(Diffuse diffuse) {
    	if(molecule == null) {
    		double[][][][] values = new double[NUM_STATES][GridFactory.getXbin()][GridFactory.getYbin()][GridFactory.getZbin()];
    		molecule = new Transferrin(values, diffuse); 
    	}
    	return molecule;
    }
    
    public static Transferrin getMolecule() {
    	return getMolecule(null);
    }
    
    @Override
    public double getKd() {
    	return Constants.Kd_TfR2;
    }
    
    public void turnOver(int x, int y, int z) {}
    
    /**
     * Disabled.
     */
    public void degrade() {}

    public void computeTotalMolecule(int x, int y, int z) {
    	this.totalMoleculesAux[0] = this.totalMoleculesAux[0] + this.get(0, x, y, z);
    	this.totalMoleculesAux[1] = this.totalMoleculesAux[1] + this.get(1, x, y, z);
    	this.totalMoleculesAux[2] = this.totalMoleculesAux[2] + this.get(2, x, y, z);
    }

    public int getIndex(String str) {
        return Transferrin.INDEXES.get(str);
    }

    protected boolean templateInteract(Interactable interactable, int x, int y, int z) {
        if (interactable instanceof Macrophage)
        	return Interactions.transferrinMacrophage((Leukocyte) interactable, this, AspergillusMacrophage._FPN, AspergillusMacrophage.M1, x, y, z);
        	
        if(interactable instanceof Iron)
        	return Interactions.transferrinIronChelation(this, (Molecule) interactable, x, y, z);
        
        return interactable.interact(this, x, y, z);
    }

	@Override
	public String getName() {
		return NAME;
	} 

	@Override
	public double getThreshold() {
		return THRESHOLD;
	}

	@Override
	public int getNumState() {
		return NUM_STATES;
	}
	
	@Override
	public boolean isTime() {
		return true;
	}
}

File: src/edu/uf/interactable/Transferrin.cpp
/*
 * Transferrin.cpp
 *
 *  Created on: Jul 19, 2024
 *      Author: henriquedeassis
 */

#include "Transferrin.h"
#include "../utils/Id.h"
#include "../utils/Constants.h"
#include "../utils/Util.h"
#include "../primitives/Interactions.h"
#include "Macrophage.h"
#include "afumigatus/TAFC.h"
#include <memory>
#include "../intracellularState/MATestModel.h"

namespace edu {
namespace uf {
namespace interactable {

using namespace primitives;
using namespace afumigatus;

const int Transferrin::classId = Id::getId();
const string Transferrin::NAME = "Transferrin";
const double Transferrin::THRESHOLD = utils::constexprants::K_M_TF_TAFC * utils::constexprants::VOXEL_VOL / 1.0e6;
Transferrin* Transferrin::molecule = nullptr;

Transferrin::Transferrin(double**** qttys, int xbin, int ybin, int zbin, char wbin, Diffuse* diffuse)
    : Molecule(qttys, xbin, ybin, zbin, wbin, diffuse) {}

Transferrin* Transferrin::getMolecule(Diffuse* diffuse, int xbin, int ybin, int zbin, char wbin) {
	if (molecule == nullptr) {
	        double**** values = new double***[wbin];
	        for (int i = 0; i < wbin; ++i) {
	            values[i] = new double**[xbin];
	            for (int j = 0; j < xbin; ++j) {
	                values[i][j] = new double*[ybin];
	                for (int k = 0; k < ybin; ++k) {
	                    values[i][j][k] = new double[zbin]();
	                }
	            }
	        }
	        molecule = new Transferrin(values, xbin, ybin, zbin, wbin, diffuse);
	    }
	    return molecule;
}

int Transferrin::getClassId() const {
	return classId;
}

double Transferrin::getKd() {
    return utils::constexprants::Kd_TfR2;
}

int Transferrin::getIndex(const char* str) const{
	return INDEXES.at(str);
}

void Transferrin::turnOver(int x, int y, int z) {}

void Transferrin::degrade() {}

void Transferrin::computeTotalMolecule(int x, int y, int z) {
	totalMoleculesAux[0] += get(0, x, y, z);
	totalMoleculesAux[1] += get(1, x, y, z);
	totalMoleculesAux[2] += get(2, x, y, z);
}

bool Transferrin::templateInteract(Interactable* interactable, int x, int y, int z) {
    /*if (auto transferrin = dynamic_cast<Transferrin*>(interactable)) {
        return Interactions::siderophoreTransferrinChelation(dynamic_cast<Molecule*>(transferrin), this, Constants::K_M_TF_TAFC, x, y, z);
    }*/
    if (auto macrophage = dynamic_cast<Macrophage*>(interactable))
        return Interactions::transferrinMacrophage(macrophage, this, MATestModel::_FPN,  MATestModel::M1, x, y, z);

    //if(auto iron = dynamic_cast<Iron*>(interactable))
    //	return Interactions::transferrinIronChelation(this, iron, x, y, z);

    if(auto tafc = dynamic_cast<TAFC*>(interactable))
    	return Interactions::siderophoreTransferrinChelation(this, tafc, utils::constexprants::K_M_TF_TAFC, x, y, z);
    /*if (auto iron = dynamic_cast<Iron*>(interactable)) {
        return Interactions::siderophoreIronChelation(dynamic_cast<Molecule*>(iron), this, x, y, z);
    }*/
    return interactable->interact(this, x, y, z);
}

string Transferrin::getName() const {
    return NAME;
}

double Transferrin::getThreshold() const {
    return THRESHOLD;
}

int Transferrin::getNumState() const {
    return NUM_STATES;
}

bool Transferrin::isTime() const {
    return true;
}

} /* namespace interactable */
} /* namespace uf */
} /* namespace edu */
